core/src/main/scala/org/apache/spark/util/UninterruptibleThread.scala

```diff
@@ -101,7 +101,11 @@ private[spark] class UninterruptibleThread(
      *         is true) and no concurrent [[interrupt()]] call, otherwise false
      */
     def isInterruptible: Boolean = synchronized {
-      shouldInterruptThread = uninterruptible
+      // SPARK-53394: We should not interrupt the thread when it is already interrupted.
+      // Otherwise, the state of `shouldInterruptThread` becomes inconsistent between
+      // `isInterruptible()` and `isInterruptPending()`, leading to `UninterruptibleThread`
+      // be interruptible under `runUninterruptibly`.
+      shouldInterruptThread = uninterruptible || UninterruptibleThread.this.isInterrupted
```

Comment:
I think `isInterrupted` can be moved to line 115:
```
if (!shouldInterruptThread && !awaitInterruptThread && !isInterrupted) {
...
```
along with updating comment "there is no other threads that called interrupt" to cover both conditions: "awaitInterruptThread is already true or isInterrupted is true.

Reply:
TBH, I think that would be less readable.